`PartialWitnessGenerator` trait is not object safe

### Problem

I'm in the process of writing the "nargo core" library which must allow passing in the backend which we want to use (for reasons discussed [here](https://github.com/noir-lang/noir/pull/1085#issuecomment-1494618926)). I'm trying to pass these as a trait object (i.e. `Box<dyn PartialWitnessGenerator>`) so the backend can be chosen in a different crate and passed in however this fails due to some methods which are making the trait non-object safe.


```
error[E0038]: the trait `acvm::Backend` cannot be made into an object
   --> crates/nargo/src/ops/execute.rs:7:18
    |
7   |     backend: Box<dyn Backend>,
    |                  ^^^^^^^^^^^ `acvm::Backend` cannot be made into an object
    |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
   --> /home/tom/.cargo/registry/src/github.com-1ecc6299db9ec823/acvm-0.8.0/src/lib.rs:162:8
    |
162 |     fn solve_black_box_function_call(
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait cannot be made into an object because associated function `solve_black_box_function_call` has no `self` parameter
...
169 |     fn any_missing_assignment(
    |        ^^^^^^^^^^^^^^^^^^^^^^ the trait cannot be made into an object because associated function `any_missing_assignment` has no `self` parameter
...
182 |     fn solve_directives(
    |        ^^^^^^^^^^^^^^^^ the trait cannot be made into an object because associated function `solve_directives` has no `self` parameter

```

### Proposed solution

We should either move these methods out of the trait (`any_missing_assignments` doesn't benefit from being a trait method rather than being a regular function) or add a `self` parameter.

### Alternatives considered

_No response_

### Additional context

_No response_

### Submission Checklist

- [X] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

--- a/acvm/src/lib.rs
+++ b/acvm/src/lib.rs
@@ -325,7 +325,8 @@ mod test {
     };
 
     use crate::{
-        pwg::block::Blocks, OpcodeResolution, OpcodeResolutionError, PartialWitnessGenerator,
+        pwg::block::Blocks, Backend, OpcodeResolution, OpcodeResolutionError,
+        PartialWitnessGenerator,
     };
 
     struct StubbedPwg;
@@ -407,5 +408,13 @@ mod test {
             .expect("should be solvable");
         assert!(unsolved_opcodes.is_empty(), "should be fully solved");
         assert!(unresolved_oracles.is_empty(), "should have no unresolved oracles");
+    }
+
+    #[test]
+    fn test_backend_object_safety() {
+        // This test doesn't do anything at runtime.
+        // We just want to ensure that the `Backend` trait is object safe and this test will refuse to compile
+        // if this property is broken.
+        fn check_object_safety(_backend: Box<dyn Backend>) {}
     }
 }

--- a/acvm/src/lib.rs
+++ b/acvm/src/lib.rs
@@ -10,14 +10,13 @@ pub mod pwg;
 use crate::pwg::{arithmetic::ArithmeticSolver, oracle::OracleSolver};
 use acir::{
     circuit::{
-        directives::Directive,
         opcodes::{BlackBoxFuncCall, OracleData},
         Circuit, Opcode,
     },
     native_types::{Expression, Witness},
     BlackBoxFunc,
 };
-use pwg::block::Blocks;
+use pwg::{block::Blocks, directives::solve_directives};
 use std::collections::BTreeMap;
 use thiserror::Error;
 
@@ -65,6 +64,22 @@ pub enum OpcodeResolution {
     InProgress,
 }
 
+/// Check if all of the inputs to the function have assignments
+///
+/// Returns the first missing assignment if any are missing
+fn first_missing_assignment(
+    witness_assignments: &BTreeMap<Witness, FieldElement>,
+    func_call: &BlackBoxFuncCall,
+) -> Option<Witness> {
+    func_call.inputs.iter().find_map(|input| {
+        if witness_assignments.contains_key(&input.witness) {
+            None
+        } else {
+            Some(input.witness)
+        }
+    })
+}
+
 pub trait Backend: SmartContract + ProofSystemCompiler + PartialWitnessGenerator {}
 
 /// This component will generate the backend specific output for
@@ -89,18 +104,16 @@ pub trait PartialWitnessGenerator {
                     Opcode::Arithmetic(expr) => ArithmeticSolver::solve(initial_witness, expr),
                     Opcode::BlackBoxFuncCall(bb_func) => {
                         if let Some(unassigned_witness) =
-                            Self::any_missing_assignment(initial_witness, bb_func)
+                            first_missing_assignment(initial_witness, bb_func)
                         {
                             Ok(OpcodeResolution::Stalled(OpcodeNotSolvable::MissingAssignment(
                                 unassigned_witness.0,
                             )))
                         } else {
-                            Self::solve_black_box_function_call(initial_witness, bb_func)
+                            self.solve_black_box_function_call(initial_witness, bb_func)
                         }
                     }
-                    Opcode::Directive(directive) => {
-                        Self::solve_directives(initial_witness, directive)
-                    }
+                    Opcode::Directive(directive) => solve_directives(initial_witness, directive),
                     Opcode::Block(block) | Opcode::ROM(block) | Opcode::RAM(block) => {
                         blocks.solve(block.id, &block.trace, initial_witness)
                     }
@@ -160,37 +173,10 @@ pub trait PartialWitnessGenerator {
     }
 
     fn solve_black_box_function_call(
+        &self,
         initial_witness: &mut BTreeMap<Witness, FieldElement>,
         func_call: &BlackBoxFuncCall,
     ) -> Result<OpcodeResolution, OpcodeResolutionError>;
-
-    // Check if all of the inputs to the function have assignments
-    // Returns the first missing assignment if any are missing
-    fn any_missing_assignment(
-        initial_witness: &mut BTreeMap<Witness, FieldElement>,
-        func_call: &BlackBoxFuncCall,
-    ) -> Option<Witness> {
-        func_call.inputs.iter().find_map(|input| {
-            if initial_witness.contains_key(&input.witness) {
-                None
-            } else {
-                Some(input.witness)
-            }
-        })
-    }
-
-    fn solve_directives(
-        initial_witness: &mut BTreeMap<Witness, FieldElement>,
-        directive: &Directive,
-    ) -> Result<OpcodeResolution, OpcodeResolutionError> {
-        match pwg::directives::solve_directives(initial_witness, directive) {
-            Ok(_) => Ok(OpcodeResolution::Solved),
-            Err(OpcodeResolutionError::OpcodeNotSolvable(unsolved)) => {
-                Ok(OpcodeResolution::Stalled(unsolved))
-            }
-            Err(err) => Err(err),
-        }
-    }
 }
 
 pub trait SmartContract {
@@ -333,6 +319,7 @@ mod test {
 
     impl PartialWitnessGenerator for StubbedPwg {
         fn solve_black_box_function_call(
+            &self,
             _initial_witness: &mut BTreeMap<Witness, FieldElement>,
             _func_call: &BlackBoxFuncCall,
         ) -> Result<OpcodeResolution, OpcodeResolutionError> {
@@ -415,6 +402,7 @@ mod test {
         // This test doesn't do anything at runtime.
         // We just want to ensure that the `Backend` trait is object safe and this test will refuse to compile
         // if this property is broken.
+        #[allow(dead_code)]
         fn check_object_safety(_backend: Box<dyn Backend>) {}
     }
 }

--- a/acvm/src/pwg/directives.rs
+++ b/acvm/src/pwg/directives.rs
@@ -8,11 +8,30 @@ use acir::{
 use num_bigint::BigUint;
 use num_traits::Zero;
 
-use crate::OpcodeResolutionError;
+use crate::{OpcodeResolution, OpcodeResolutionError};
 
 use super::{get_value, insert_value, sorting::route, witness_to_value};
 
+/// Attempts to solve the [`Directive`] opcode `directive`.
+/// If successful, `initial_witness` will be mutated to contain the new witness assignment.
+///
+/// Returns `Ok(OpcodeResolution)` to signal whether the directive was successful solved.
+///
+/// Returns `Err(OpcodeResolutionError)` if a circuit constraint is unsatisfied.
 pub fn solve_directives(
+    initial_witness: &mut BTreeMap<Witness, FieldElement>,
+    directive: &Directive,
+) -> Result<OpcodeResolution, OpcodeResolutionError> {
+    match solve_directives_internal(initial_witness, directive) {
+        Ok(_) => Ok(OpcodeResolution::Solved),
+        Err(OpcodeResolutionError::OpcodeNotSolvable(unsolved)) => {
+            Ok(OpcodeResolution::Stalled(unsolved))
+        }
+        Err(err) => Err(err),
+    }
+}
+
+fn solve_directives_internal(
     initial_witness: &mut BTreeMap<Witness, FieldElement>,
     directive: &Directive,
 ) -> Result<(), OpcodeResolutionError> {

